core/modules/loader/MWorkerList.cc

```diff
@@ -0,0 +1,424 @@
+                MWorkerListItem::Ptr const& wListItem = item.second;
+                protoItem->set_wid(wListItem->getId());
+            }
+            protoList.SerializeToString(&(workerList.element));
+            LoaderMsg workerListMsg(LoaderMsg::MAST_WORKER_LIST, msgId, ourHostName, ourPort);
+            _stateListData = std::make_shared<BufferUdp>();
+            workerListMsg.appendToData(*_stateListData);
+            workerList.appendToData(*_stateListData);
+        }
+    }
+
+    _central->sendBufferTo(ip, port, *_stateListData);
+
+    // See if this worker is know.
+    MWorkerListItem::Ptr workerItem;
+    {
+        // See if this is a worker in our map
+        std::lock_guard<std::mutex> lock(_mapMtx);
+        auto iter = _ipMap.find(address);
+        if (iter != _ipMap.end()) {
+            workerItem = iter->second;
+        }
+    }
+    if (workerItem != nullptr) {
+        workerItem->sendListToWorkerInfoReceived();
+    }
+    return true;
+}
+
+
+
+
+std::pair<std::vector<MWorkerListItem::Ptr>, std::vector<MWorkerListItem::Ptr>>
+MWorkerList::getActiveInactiveWorkerLists() {
+    std::vector<MWorkerListItem::Ptr> active;
+    std::vector<MWorkerListItem::Ptr> inactive;
+    std::lock_guard<std::mutex> lck(_mapMtx);
+    for(auto const& elem : _nameMap) {
+        auto item = elem.second;
+        if (item->isActive()) {
+            active.push_back(item);
+        } else {
+            inactive.push_back(item);
+        }
+    }
+    auto pair = std::make_pair(active, inactive);
+    return pair;
+}
+
+
+// must lock _mapMtx before calling this function
+void MWorkerList::_flagListChange() {
+    _wListChanged = true;
+    // On the Master, flag each worker in the list that it needs to send an updated list to it's worker.
+    for (auto const& elem : _nameMap) {
+        auto const& item = elem.second;
+        item->flagNeedToSendList();
+    }
+}
+
+
+std::string MWorkerList::dump() const {
+    std::stringstream os;
+    os << "MWorkerList:\n";
+    {
+        std::lock_guard<std::mutex> lck(_mapMtx);
+        for (auto elem:_nameMap) {
+            os << "  " << *elem.second << "\n";
+        }
+        os << "MWorkerList ip:\n";
+        for (auto elem:_ipMap) {
+            os << "  " << *elem.second << "\n";
+        }
+    }
+    return os.str();
+}
+
+
+void MWorkerListItem::addDoListItems(Central *central) {
+    LOGS(_log, LOG_LVL_DEBUG, "MWorkerListItem::addDoListItems");
+    std::lock_guard<std::mutex> lck(_doListItemsMtx);
+    if (_sendListToWorker == nullptr) {
+        _sendListToWorker = std::make_shared<SendListToWorker>(shared_from_this(), _central);
+        _central->addDoListItem(_sendListToWorker);
+    }
+    if (_reqWorkerKeyInfo == nullptr) {
+        _reqWorkerKeyInfo = std::make_shared<ReqWorkerKeyInfo>(shared_from_this(), _central);
+        _central->addDoListItem(_reqWorkerKeyInfo);
+    }
+}
+
+
+void MWorkerListItem::flagNeedToSendList() {
+    auto slw = _sendListToWorker;
```

Comment:
Is this thread safe?

Reply:
Making a copy of the shared_ptr is a way to make sure the shared_ptr is not modified (reset, or reassigned) while it is in use. As far as I can tell the pointer object _is_ threadsafe (while it does not make any guarantees about the pointed-to object) so this should be thread safe.

Reply:
However, I can't tell that this copying is necessary, since once an object is assigned to the ptr, that ptr is never changed, so unless I'm missing something I think you could just say `if (_sendListToWorker != nullptr) { _sendListToWorker->setNeedInfo(); }`

Reply:
This is just a check to make sure there isn't a segfault.
Once _sendListToWorker has been set, it shouldn't be changed unless this entire MWorkerListItem is removed and all calls like this become moot. 
I don't thing there's a way to make this obvious in the code but I'll add some comments.